### ISSUE TITLE
RAM not cleared even after moving stable diffusion pipeline to cuda

### ISSUE THREAD

**Issue by hari10599** [huggingface/diffusers#3223]
### Describe the bug

Moving the SD model to cuda after loading it, doesn't clear the RAM.

Before loading SD model,
- GPU usage:  0MiB / 49140MiB 
- RAM usage: 2.232330322265625/41.46642303466797

After loading SD model,
- GPU usage: 3755MiB / 49140MiB
- RAM usage: 5.694705963134766/37.9942741394043

Why isn't the RAM cleared even after moving the model to GPU?

### Reproduction

```
import torch
from diffusers import StableDiffusionDepth2ImgPipeline
from diffusers import UniPCMultistepScheduler
from PIL import Image

pipe = StableDiffusionDepth2ImgPipeline.from_pretrained(
    'stabilityai/stable-diffusion-2-depth',
    torch_dtype=torch.float16
)
pipe.scheduler = UniPCMultistepScheduler.from_config(pipe.scheduler.config)
pipe.to('cuda')
pipe.enable_xformers_memory_efficient_attention()
```

### System Info

accelerate 0.18.0
bitsandbytes 0.38.1
diffusers 0.15.1

**Comment by LLjo** [comment]
I'm struggling with RAM issues as well. Im not quite sure if its a WSL problem, are you using it as well? When switching models without restarting my fastapi service no RAM ever gets released, using gc.collect().

**Comment by init-22** [comment]
@hari10599 @LLjo Usually when converting the model to GPU PyTorch makes a copy of the loaded cpu model and then you have to delete the previous model (using del model or model.__del__()) 

I think if we try to store pipe.to('cuda') in new variable and delete the pipe then ideally we should be able to see reduced size in RAM allocation but since the model is not just a PyTorch model but an instance of StableDiffusionPipeline it won't directly free up the memory unless you implement a __del__() method  to free up everything in the pipeline 


@patrickvonplaten @yiyixuxu @sayakpaul  correct me if I'm wrong on the above points and it there any suggestion on this?

**Comment by hari10599** [comment]
> @hari10599 @LLjo Usually when converting the model to GPU PyTorch makes a copy of the loaded cpu model and then you have to delete the previous model (using del model or model.**del**())
> 
> I think if we try to store pipe.to('cuda') in new variable and delete the pipe then ideally we should be able to see reduced size in RAM allocation but since the model is not just a PyTorch model but an instance of StableDiffusionPipeline it won't directly free up the memory unless you implement a **del**() method to free up everything in the pipeline
> 
> @patrickvonplaten @yiyixuxu @sayakpaul correct me if I'm wrong on the above points and is there any suggestion on this?

Is there any recommended way to remove a specific model/pipeline from RAM/GPU? @patil-suraj @patrickvonplaten

**Comment by sayakpaul** [comment]
If your pipeline does not need a particular model you can do `del your_model` to delete the Python object referencing allocated GPU memory.

Additionally, we also recommend:

```python
import gc
import torch

def flush():
  gc.collect()
  torch.cuda.empty_cache()
```

**Comment by LLjo** [comment]
> @hari10599 @LLjo Usually when converting the model to GPU PyTorch makes a copy of the loaded cpu model and then you have to delete the previous model (using del model or model.**del**())
> 
> I think if we try to store pipe.to('cuda') in new variable and delete the pipe then ideally we should be able to see reduced size in RAM allocation but since the model is not just a PyTorch model but an instance of StableDiffusionPipeline it won't directly free up the memory unless you implement a **del**() method to free up everything in the pipeline
> 
> @patrickvonplaten @yiyixuxu @sayakpaul correct me if I'm wrong on the above points and is there any suggestion on this?

Yep del is being used on the pipes created. I need to test loading in several models with a pure python script and then try free up the memory. I'm suspecting it could be some other part playing a role here like FastAPI.

**Comment by init-22** [comment]
@LLjo Okay, but I couldn't find any reference to FastAPI so it's mostly likely not the case

**Comment by LLjo** [comment]
All right so it seems like it's just WSL being bad as usually, running this doesnt clear any RAM in WSL except for cuda memory:

`from diffusers import DiffusionPipeline
import torch
import time
import gc

def test():
    pipeline = DiffusionPipeline.from_pretrained("stable-diffusion-v1-5").to("cuda")

    pipeline2 = DiffusionPipeline.from_pretrained("stable-diffusion-v1-5").to("cuda")
    
    print('del pipes')
    
    del pipeline
    del pipeline2
    gc.collect()
    torch.cuda.empty_cache()
    
    print('done, check ram')
    time.sleep(20)

test()`

**Comment by LLjo** [comment]
I had a diffusers setup in windows before without WSL which worked fine to swap models. @hari10599 Are you using WSL? If so then we can probably close this

**Comment by hari10599** [comment]
> I had a diffusers setup in windows before without WSL which worked fine to swap models. @hari10599 Are you using WSL? If so then we can probably close this

No. I'm using a notebook in Paperspace gradient
